### PR TITLE
Refactor network delegate helpers to not have access to the request

### PR DESCRIPTION
**Pull request by bbondy**
Fix: https://github.com/brave/brave-browser/issues/1574
Fix: https://github.com/brave/brave-browser/issues/1576

Also don't call Cancel() before a request is started

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Windows
  - [x] macOS
  - [ ] Linux
- Verified that these changes pass automated tests (`npm test brave_unit_tests && npm test brave_browser_tests`) on
  - [ ] Windows
  - [x] macOS
  - [ ] Linux
- [ ] Ran `git rebase master` (if needed).
- [ ] Ran `git rebase -i` to squash commits (if needed).
- [ ] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [ ] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

### CODE REVIEW

File: browser/net/brave_ad_block_tp_network_delegate_helper.cc [hunk 1]
```diff
@@ -91,93 +89,77 @@ bool GetPolyfillForAdBlock(bool allow_brave_shields, bool allow_ads,
...
-void OnBeforeURLRequest_OnBeforeURLRequest_AdBlockTPPostCheckWork(
-    net::URLRequest* request,
+void OnBeforeURLRequestDispatchOnIOThread(
     GURL* new_url,
     const ResponseCallback& next_callback,
     std::shared_ptr<BraveRequestInfo> ctx) {
   DCHECK_CURRENTLY_ON(content::BrowserThread::IO);
   if (!ctx->new_url_spec.empty() &&
-    ctx->new_url_spec != request->url().spec()) {
+    ctx->new_url_spec != ctx->request_url.spec()) {
```
Comment (bridiver) [inline browser/net/brave_ad_block_tp_network_delegate_helper.cc]: follow up we should look at not passing `new_url` to the callbacks and instead having a const ctx.original_request_url and let the callbacks modify ctx.request_url
in the base delegate we can set new_url if request_url != original_request_url

File: browser/net/brave_ad_block_tp_network_delegate_helper.cc [hunk 2]
```diff
@@ -91,93 +89,77 @@ bool GetPolyfillForAdBlock(bool allow_brave_shields, bool allow_ads,
...
-    net::URLRequest* request,
+int OnBeforeURLRequest_AdBlockTPPreWork(
     GURL* new_url,
     const ResponseCallback& next_callback,
     std::shared_ptr<BraveRequestInfo> ctx) {
-  const GURL& url = request->url();
-  GURL tab_origin = request->site_for_cookies().GetOrigin();
 
-  if (request->url().is_empty()) {
+  if (ctx->request_url.is_empty()) {
```
Comment (bridiver) [inline browser/net/brave_ad_block_tp_network_delegate_helper.cc]: see above about assuming new_url hasn't been set yet

Comment (bridiver) [reply browser/net/brave_ad_block_tp_network_delegate_helper.cc]: but not a problem introduced by this change

File: browser/net/brave_ad_block_tp_network_delegate_helper.cc (hunk 1 above)
Comment (bbondy) [reply browser/net/brave_ad_block_tp_network_delegate_helper.cc]: posted here for the follow-up:
https://github.com/brave/brave-browser/issues/1576

**Comment by bbondy** [comment]
master: 2524553c76f06e156a658252862c87f2e2fd1d9c
0.56.x: 2c65c46241035c91315553e62910ba5c46f7ada2
0.55.x: ca0c92cf7880fc04d3ab4ec104fa6d45499c720b

**Pull request merged**